### ISSUE TITLE
[discuss] sbt auto-imports should be constrained to a single package-object

### ISSUE THREAD

**Issue by timcharper** [sbt/sbt#2447]
I'd like sbt `1.0.0` to either encourage or force sbt plugin authors to only auto-import a single package-object into namespace, for the following reasons:
- **Improved build.sbt legibility**: Generally, I avoid `import package._` because it hurts code legibility. Presently, it's difficult to tell from which plugins a particular task or SettingKey comes
  
  Instead of this:
  
  ``` scala
  Revolver.settings
  
  PB.protobufSettings
  
  enablePlugins(JavaServerAppPackaging)
  
  maintainer := "me <me@email.com>"
  ```
  
  I'd much prefer to read this:
  
  ``` scala
  revolver.defaults
  
  scalapb.defaults
  
  enablePlugins(native.JavaServerAppPackaging)
  
  native.maintainer := "me <me@email.com>"
  
  ...
  ```

...
- **Reduced symbol collisions**: The probability that the symbols from two separate sbt plugins never collide is presently too high.

(related, while making this change, we could add some text to the [Plugins Best Practices](http://www.scala-sbt.org/0.13/docs/Plugins-Best-Practices.html) guide, suggesting names for common things such as "your default settings should be {package_name}.defaults".

**Comment by dwijnand** [comment]
What would be the impact of this with regards to if different plugins provide their own keys for the same key name? Such as `native.maintainer` and `sonatype.maintainer`?

**Comment by timcharper** [comment]
In summary, the impact of this is it would solve a present issue involving namespace collisions.

tl;dr

To my knowledge, @dwijnand, [sbt-sonatype](https://github.com/xerial/sbt-sonatype) defines no such `maintainer` key. However, [sbt-native-project does right here](https://github.com/sbt/sbt-native-packager/blob/a38b34f/src/main/scala/com/typesafe/sbt/packager/Keys.scala#L14).

When two independent plugins define an overlapping key, the result is an error. To reproduce for yourself, put this in `~/.sbt/0.13/plugins/luchadorer.scala`

``` scala
package com.sbt.llave_luchadores

import sbt._

object LuchadorUno extends AutoPlugin {
  object autoImport {
    val conflictador = settingKey[String]("Tan innocente")
  }
}

object LuchadorDos extends AutoPlugin {
  object autoImport {
    val conflictador = settingKey[String]("Mucho conflicto")
  }
}
```

Then, in some project, put this in `build.sbt`:

```
conflictador := "HI"
```

This is the output:

```
timcharper@tim-scimmia:~/src/goof/test $ sbt
[info] Loading global plugins from /Users/timcharper/.sbt/0.13/plugins
[info] Compiling 1 Scala source to /Users/timcharper/.sbt/0.13/plugins/target/scala-2.10/sbt-0.13/classes...
[info] Loading project definition from /Users/timcharper/src/goof/test/project
The s3:// URLStreamHandler is already installed
Using the existing Ivy URLHandlerDispatcher to handle s3:// URLs
/Users/timcharper/src/goof/test/build.sbt:4: error: reference to conflictador is ambiguous;
it is imported twice in the same scope by
import _root_.com.sbt.llave_luchadores.LuchadorDos.autoImport._
and import _root_.com.sbt.llave_luchadores.LuchadorUno.autoImport._
conflictador := "HI"
^
```

**Comment by dwijnand** [comment]
Yeah, which makes sense. If you namespace the key values but don't namespace the key names (aka `nativeMaintainer` and `sonatypeMaintainer`) then you end up with confusion:

``` scala
LuchadorUno.autoImport.conflictador := "HI-1"
LuchadorDos.autoImport.conflictador := "HI-2"
```

```
> show conflictador
[info] HI-2
```

**Comment by eed3si9n** [comment]
The topic of plugin key namespace has been discussed several times. One such instance is https://groups.google.com/d/msg/simple-build-tool/cCCYbopR6JQ/SS7gCRLdRXAJ from 2011. The knowledge from the discussion resulted to the sbt plugin best practice guide http://www.scala-sbt.org/0.13/docs/Plugins-Best-Practices.html, and it has influenced the sbt 0.13 DSL.

Each plugin should prefix their setting keys appropriately and there's not much else we can do.
The reason is because the key-value pair only looks at the key name, not the Scala instance name. Unless we drastically change the way sbt shell is implemented, that's what we have to deal with.

**Comment by timcharper** [comment]
@eed3si9n how would you feel if an SBT AutoPlugin was required to have some prefix, and warnings were issued for all custom defined TaskKey / SettingKey's defined in the auto import space that weren't prefixed by that?

As I was developing `sbt-jnaerator`, I forced myself to prefix custom keys, and it led me to use less keys (and instead use a data-structure describing my build).

There could be an annotation to turn of the warning. I think sbt-native-packager, in particular, has a pretty good reason to not abide that convention. But for other sbt plugins, such as ensime-sbt, sbt-protobuf, etc., I don't see as much of a reason to do it.

**Comment by eed3si9n** [comment]
If it's possible to implement such warning just by looking at the keys, and not require additional `val prefix = "foo"`, I'd be open to a pull request.

**Comment by eed3si9n** [comment]
I am closing this due to inactivity, but feel free to reopen in https://discuss.lightbend.com/c/tooling.
